### ISSUE TITLE
Middle scanline looks wrong in SDL renderer (on Mac)

### ISSUE THREAD

**Issue by sdsnatcher73** [openMSX/openMSX#1166]
When starting openMSX wit scale 2 or 3 and scaler simple on recent builds there is a line halfway top to bottom that is drawn incorrectly (color seems darker). If within that session I switch to scale factor 2 or 1 and even back to 3 the off line is gone. With scale factor 2 the entire image appears too dark though. If I close openMSX (so scale_factor will be 2 in settings.xml) and open it again, the image is still too dark and the 1 line in the middle is bright (or good). Different scalers also seem to have issues: TV, RGBtriplet. I've taken screenshots of all 3 scalers with factor 2 and 3.
<img width="708" alt="RGBtriplet_scale_2" src="https://user-images.githubusercontent.com/26295777/59708441-136e5380-9205-11e9-9547-f3d68e22a649.png">
<img width="1028" alt="RGBtriplet_scale_3" src="https://user-images.githubusercontent.com/26295777/59708444-15d0ad80-9205-11e9-8f05-954450a923af.png">
<img width="708" alt="simple_scale_2" src="https://user-images.githubusercontent.com/26295777/59708445-15d0ad80-9205-11e9-8ed7-a2c13f0ab304.png">
<img width="1028" alt="simple_scale_3" src="https://user-images.githubusercontent.com/26295777/59708447-15d0ad80-9205-11e9-9ce7-5f0f24e0ae44.png">
<img width="708" alt="TV_scale_2" src="https://user-images.githubusercontent.com/26295777/59708448-15d0ad80-9205-11e9-8e97-c3e0c9a88582.png">
<img width="1028" alt="TV_scale_3" src="https://user-images.githubusercontent.com/26295777/59708449-16694400-9205-11e9-8bae-5095654e7907.png">

**Comment by MBilderbeek** [comment]
What happens if you use the SDLGL-PP render?

**Comment by sdsnatcher73** [comment]
It's not happening with the SDLGL-PP renderer... Trying to figure out the difference but I am having trouble accessing documentation on openmsx.org (which might be on my end, not sure yet)...

**Comment by MBilderbeek** [comment]
The difference is explained in chapter 6.1 of the User Manual: https://openmsx.org/manual/user.html#renderers

The bug is in the SDL renderer. I also noticed something else: clipping is not working. Check it when you open the menu in the SDL renderer: the version number is not clipped inside the menu area.

**Comment by sdsnatcher73** [comment]
Okay, is that openMSX SDL code or a bug in the library?

**Comment by MBilderbeek** [comment]
That will require further analysis. Unfortunately, I'm not very familiar with the openMSX video code.

I couldn't reproduce it on Linux. Can you reproduce it on other systems?

**Comment by sdsnatcher73** [comment]
It's fine on Windows. I'll try to build from source both master and original 0.15.0 (this should build a 0.15.0 with the latest sdl libs) if that 0.15.0 shows the same issue than it's probably in the sdl library (and/or how it is called by openMSX)

**Comment by sdsnatcher73** [comment]
Okay I rebuild 0.15.0 and SDL is completely broken on MacOS for this build. SDLGL-PP works fine, SDL gives a completely white screen... I have been building on MacOS Mojave 10.14.5 using Command Line Tools for Xcode 10.2 (which is not supported according to Compilation Guide but the build is fine and I was unable to complete the build on Xcode 9.4.1).

I also noted that building 324 introduced 2 other renderers which are not documented yet SDLGL-FB16 and SDLGL-FB32 (which are also visible in Windows 324 build)

**Comment by MBilderbeek** [comment]
On Windows you tried both SDL and SDLGL-PP renderers?

One of the reasons we moved to SDL2 is that on Mac with modern OS, the SDL renderer isn't working correctly with the SDL1 library. 

The other renderers are experimental and normally disabled. You can ignore them.

**Comment by sdsnatcher73** [comment]
Yes for me SDL and SDLGL-PP work fine on Windows 10 1904. I only see the issue on MacOS with SDL

**Comment by sdsnatcher73** [comment]
Actually it seems there is another issue with scaliness that also affects SDLGL-PP on MacOS. It happens when entering full screen... The scanlines are not distributed evenly...
<img width="1280" alt="Screenshot 2019-06-19 at 21 05 11" src="https://user-images.githubusercontent.com/26295777/59793391-b93dc300-92d6-11e9-89b4-813ede2c5da4.png">

**Comment by MBilderbeek** [comment]
Which scale factor is that?

**Comment by MBilderbeek** [comment]
By the way, for some reason, you removed the settings you were using from the initial post.
I think it was this:
`<!DOCTYPE settings SYSTEM 'settings.dtd'> <settings> <settings> 
<setting id="invalid_psg_directions_callback">psgdirectioncallback</setting> 
<setting id="di_halt_callback">dihaltcallback</setting>
<setting id="default_machine">National_CF-3000</setting> 
<setting id="deinterlace">false</setting> <setting id="blur">0</setting> 
<setting id="renderer">SDL</setting> 
<setting id="consolebackground">skins/ConsoleBackgroundBlue.png</setting> 
<setting id="horizontal_stretch">272.0</setting> 
<setting id="scanline">100</setting>
<setting id="scale_factor">3</setting> </settings> <bindings/> </settings> `

So, scanline setting is 100% and scale_factor is 3. Could be relevant.

**Comment by MBilderbeek** [comment]
> I also noticed something else: clipping is not working. Check it when you open the menu in the SDL renderer: the version number is not clipped inside the menu area.

Wouter (@m9710797) fixed that today in b76267877926633079da80b6ec65144fbe86a104

**Comment by sdsnatcher73** [comment]
Found something that improves the situation for SDL renderer (not sure if it is perfect yet) but it definitely fixes the scanlines in SDLGL-PP (where they are not drawn evenly distributed). It was described here https://stackoverflow.com/questions/18544881/sdl-2-0-retina-mac. Basically set a flag during window creation to allow High DPI. I don't have a Windows build environment (nor Linux) so can't test this on those platforms.

```
void VisibleSurface::createSurface(int width, int height, unsigned flags)
{
	if (getDisplay().getRenderSettings().getFullScreen()) {
		flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
	}

	flags |= SDL_WINDOW_ALLOW_HIGHDPI;

	assert(!window);
...
}
```

**Comment by MBilderbeek** [comment]
Can you try with different scale_factors (1-4) and see what effect this has on full screen and normal windows?

Also, what about doing it only on full screen?

So:

```
void VisibleSurface::createSurface(int width, int height, unsigned flags)
{
	if (getDisplay().getRenderSettings().getFullScreen()) {
		flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
	        flags |= SDL_WINDOW_ALLOW_HIGHDPI;
	}

	assert(!window);
...
}
```

**Comment by mthuurne** [comment]
The full screen mode we use (`SDL_WINDOW_FULLSCREEN_DESKTOP`) is basically a borderless screen-filling window. So I don't think it would make sense to only enable hi-DPI mode for full screen; just enable it always.

**Comment by sdsnatcher73** [comment]
> The full screen mode we use (`SDL_WINDOW_FULLSCREEN_DESKTOP`) is basically a borderless screen-filling window. So I don't think it would make sense to only enable hi-DPI mode for full screen; just enable it always.

I agree. I read a bit on the SDL2 bug report that lead to the implementation here: https://bugzilla.libsdl.org/show_bug.cgi?id=1934. I have no SDL2 experience so it does not make a lot of sense but they seem to be discussing the need of a hint and the use of SDL_GL_GetDrawableSize, so it may be needed that some more experienced SDL(2) developer have a look at this...

**Comment by MBilderbeek** [comment]
@peterhanraets So, can you thoroughly test with the flag as you suggested it?
Try all these combinations:
- SDL and SDLGL-PP renderers
- both on all scale_factors (1-3 for SDL and 1-4 for SDLGL-PP)
- and all of that both in windowed and full screen mode

Please let us know about the results! Thanks.

**Comment by sdsnatcher73** [comment]
Okay will do. I am also setting up MinGW on Windows and try build a Windows version with the flag set and see what happens. Windows 10 has high DPI support as well.

Do you have any idea why my builds on MacOS have issues in the console? I cannot type anything there normally, if I hold a key down it ultimately appears in the console... The builds on openmsx.dev are fine in this respect. I was wondering if it may be how the staticbindist build handles dependencies, will it fetch new versions of the dependencies if an older one already exists in 3rdparty/src. What I mean is let's say SDL2-2.0.9 exists in my 3rdparty/src folder (as it does now). I git pull the latest commits but in the meantime SDL2-2.1.0 has been released. Will it get that and build it?

Basically what it boils down to, can I reuse an existing clone that staticbindist has been made in (I can of course the real question is will the result of a new make be the same as if I would do a fresh clone)

**Comment by MBilderbeek** [comment]
I have no idea about those issues.

The staticbindist just fetches the exact libraries as are specified by ourselves. So, always the same (unless you modified something). See the build/packages.py file.

You can use your existing clone. Whatever you did with it has no effect, i.e. if `git status` doesn't list any modifications. You can make any kind of build with the same clone and they will be all reproducible.

**Comment by sdsnatcher73** [comment]
Okay, understood. Still odd that console is acting up on my build. Also in the osd menu, when switching, e.g. scale from 2 to 3 by pressing right it will go to 4 (so it fires 2 events).

Anyway back on topic. This fixes all mentioned scanline issues on MacOS when using SDLGL-PP. scanlines are distributed evenly for scale factor 2, 3 and 4 in windowed and full screen. I did however notice that scaliness don't work if scale factor is 1. Probably (I'm guessing) because at scale factor 1 the scaler is actually disabled and it is responsible for scanline generation as well). If that assumption is true enabling the scaler even at scale factor 1 should allow scanlines even then.

Now back to SDL renderer. With this flag distribution of scanlines is good. There are still some oddities (too bright or dark line), but only after openMSX is started with SDL and scale factor >1. If you toggle full screen or switch scale factor (which causes SDL window to be recreated, right?) these oddities disappear. This behaviour was already observable before I created the issue. Let's observe in the screenshots (all SDL, scale factor 2):

openMSX starts (with 2 too bright lines):
<img width="752" alt="Screenshot 2019-06-22 at 19 48 57" src="https://user-images.githubusercontent.com/26295777/59967308-75da9300-9528-11e9-9377-c0c33b597d15.png">

Switch to full screen (this is perfect imho):
<img width="1280" alt="Screenshot 2019-06-22 at 19 49 07" src="https://user-images.githubusercontent.com/26295777/59967320-973b7f00-9528-11e9-8877-c543186fd17a.png">

Switch back to windowed, also perfect:
<img width="752" alt="Screenshot 2019-06-22 at 19 49 21" src="https://user-images.githubusercontent.com/26295777/59967332-b76b3e00-9528-11e9-87d6-27250bb4995b.png">

**Comment by sdsnatcher73** [comment]
So I guess for SDL the recreated window is not 100% the same as the window that is created initially (but I could not figure out what the difference would be by going through the code)

**Comment by MBilderbeek** [comment]
When the artifact is visible, can you dump the settings to a file like this: `save_settings /tmp/mysettings.xml` and attach that file?

I tried myself (with scanline 100, scale_factor 2, SDL renderer), but I can't reproduce it:
![Screenshot from 2019-06-22 20-45-26](https://user-images.githubusercontent.com/12675978/59967738-d40a7480-952e-11e9-843f-333180a9f42d.png)

It is correct that at scale_factor 1, there are no scanlines. There are simply no pixels to make scanlines on a 320x240 area.

About the other issue which is specific to your build: better make a separate ticket for that.

**Comment by MBilderbeek** [comment]
I'll commit the change. Can you then test it on Windows 10, using the same procedure?

**Comment by sdsnatcher73** [comment]
Okay build 330 on Windows behaves perfectly in SDL and SDLGL-PP all scale factors both windowed and full screen. I'll go on to MacOS now...

**Comment by sdsnatcher73** [comment]
Build 330 from openmsx.dev behaves like my previous build. Only issue remains is SDL renderer on startup. Here are the settings:
```
<!DOCTYPE settings SYSTEM 'settings.dtd'>
<settings>
  <settings>
    <setting id="invalid_psg_directions_callback">psgdirectioncallback</setting>
    <setting id="di_halt_callback">dihaltcallback</setting>
    <setting id="default_machine">Panasonic_FS-A1ST</setting>
    <setting id="deinterlace">false</setting>
    <setting id="blur">0</setting>
    <setting id="consolebackground">skins/ConsoleBackgroundBlue.png</setting>
    <setting id="horizontal_stretch">272.0</setting>
    <setting id="scanline">100</setting>
    <setting id="osd_disk_path">/Users/peter/Sources/SD/MSX/GAME/MICROCAB</setting>
    <setting id="renderer">SDL</setting>
  </settings>
  <bindings/>
</settings>
```

The 330 build on openmsx.dev also shows the issue with switching scale factors in OSD menu (where a right press in the menu will go up twice, e.g. from 2 to 4 instead of 3). I will create an issue for this and another for the console issue on my personal build.

**Comment by MBilderbeek** [comment]
Even with these settings I can't reproduce it. Apparently you also can't reproduce it on Windows. (I also tried with Wine and my local MinGW build: no success).

So, this seems to be a macOS specific issue. I have no idea how to continue with it.

**Comment by sdsnatcher73** [comment]
I agree it is related to MacOS and most likely due to how it handles HighDPI / Retina. I think we could debug the remaining issue if we dumped the SDL window properties on initial start and after a change in a running session (like changing scale factor or switching to full screen and back to the same setting as on initial startup) and see what the difference is between initial and final window (but I would not know how to do it).

**Comment by grauw** [comment]
I tried to reproduce but I can not.

I started with default settings `renderer SDL`, `scanline 100`, `scale_factor 2` and `blur 0`. Using XCode 11.3.1 and MacOS Catalina 10.15.1 on a MacBook Pro (16-inch, 2019).

**Comment by MBilderbeek** [comment]
@peterhanraets Can you still reproduce this issue(s) yourself with the latest code/tools?

**Comment by sdsnatcher73** [comment]
This is not happening for me anymore

**Comment by MBilderbeek** [comment]
OK, as no one can reproduce it, I'm closing it.
